permits/api.py

```diff
@@ -0,0 +1,186 @@
+        """
+        This view should return a list of events for which the logged user has
+        view permissions
+        """
+        user = self.request.user
+        starts_at = self.request.query_params.get("starts_at", None)
+        ends_at = self.request.query_params.get("ends_at", None)
+        administrative_entity = self.request.query_params.get("adminentity", None)
+
+        base_filter = Q()
+        if starts_at:
+            start = datetime.datetime.strptime(starts_at, "%Y-%m-%d")
+            base_filter &= Q(starts_at__gte=start)
+        if ends_at:
+            end = datetime.datetime.strptime(ends_at, "%Y-%m-%d")
+            base_filter &= Q(ends_at__lte=end)
+        if administrative_entity:
+            base_filter &= Q(
+                permit_request__administrative_entity=administrative_entity
+            )
+
+        works_object_types_prefetch = Prefetch(
+            "permit_request__works_object_types",
+            queryset=models.WorksObjectType.objects.select_related("works_type"),
+        )
+
+        qs = (
+            models.PermitRequestGeoTime.objects.filter(base_filter)
+            .filter(
+                Q(permit_request__in=services.get_permit_requests_list_for_user(user))
+                | Q(permit_request__is_public=True)
+            )
+            .prefetch_related(works_object_types_prefetch)
+            .select_related("permit_request__administrative_entity")
+        )
+
+        return qs.order_by("starts_at")
+
+
+# //////////////////////////////////
+# PERMIT REQUEST ENDPOINT
+# //////////////////////////////////
+
+
+class BlockRequesterUserPermission(BasePermission):
+    """
+    Block access to Permit Requesters (General Public)
+    """
+
+    def has_permission(self, request, view):
+        return request.user.get_all_permissions()
+
+
+class PermitRequestViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    Permit request endpoint Usage:
+        1.- /rest/permits/?permit_request_id=1
+        2.- /rest/permits/?works_object_type=1
+        3.- /rest/permits/?status=0
+        4.- /rest/permits/?geom_type=lines | points | polygons
+    """
+
+    serializer_class = serializers.PermitRequestPrintSerializer
+    permission_classes = [IsAuthenticated, BlockRequesterUserPermission]
+
+    def get_queryset(self):
+        """
+        This view should return a list of permits for which the logged user has
+        view permissions
+        """
+        user = self.request.user
+
+        works_object_type = self.request.query_params.get("works_object_type", None)
+        status = self.request.query_params.get("status", None)
+        geom_type = self.request.query_params.get("geom_type", None)
+        permit_request_id = self.request.query_params.get("permit_request_id", None)
+
+        base_filter = Q()
+
+        if works_object_type:
+            if works_object_type.isdigit():
+                base_filter &= Q(works_object_types=works_object_type)
```

Comment:
I cannot see the previous comment, but here we are filtering the PermitRequest queryset before sending the result to the serializer, so the type checking was still left, because a TypeError might rise before validating via serializer.is_valid()

Instead of Rising an error 500, now it raises a 400 with the appropriate message.

Reply:
Merci bcp!